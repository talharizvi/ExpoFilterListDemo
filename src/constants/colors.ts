export default {
    primary: '#ffffff',
    secondary: '#dadee6',
    selectedOptionColor:'#000000',
    background: 'rgba(0, 122, 255, 0.12)',
    iconText: '#5280eb',
    userText:'#5e5e5d',
    loader:'#007bff'
  };