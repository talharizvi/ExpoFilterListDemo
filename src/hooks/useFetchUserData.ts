import React, { useEffect, useState } from "react";
import { USERS_API_URL } from "../constants/constant";

interface User {
    name: string;
    type: number;
}

interface FetchUserDataResult {
    listItems: User[];
    isLoading: boolean;
}

const useFetchUserData = ():FetchUserDataResult => {
   const [listItems, setListItems] = useState<User[]>([]);
   const [isLoading, setIsLoading] = useState<boolean>(true);

    useEffect(() => {
        const fetchData = async () => {
            try {
                const response = await fetch(USERS_API_URL);
                if (!response.ok) {
                    throw new Error(`HTTP error! Status: ${response.status}`);
                }
                const result = await response.json();
                setListItems(result?.users || []);
                setIsLoading(false);
            } catch (error) {
                console.error('Error fetching data:', error);
                setIsLoading(false);
            }
        };

        fetchData();
        // async function fetchData(){
        //     let response = await fetch(USERS_API_URL)
        //     let result  = await response.json()
        //     setListItems(result?.users || [])    
        // }

        // try {
        //     fetchData()
        // } catch (error) {
        //     console.log(error)
        // } finally {
        //     setIsLoading(false);
        // }
    }, [])

   return { listItems, isLoading };
}

export { useFetchUserData }