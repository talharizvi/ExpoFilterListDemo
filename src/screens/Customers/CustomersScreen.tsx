import React, { useState, useEffect, useCallback} from 'react';
import { StatusBar } from 'expo-status-bar';
import { StyleSheet, Text, View, ScrollView, ActivityIndicator } from 'react-native';
import { useFetchUserData } from '../../hooks/useFetchUserData';
import UserListItem from '../../components/UserListItem/UserListItem';
import CategoryHeader from '../../components/CategoryHeader/CategoryHeader';
import CustomRadioButton from '../../components/CustomRadioButton/CustomRadioButton';
import LineSeparator from '../../components/LineSeparator/LineSeparator';
import colors from '../../constants/colors';
import { filterUsersByType } from '../../utils/utils';


interface User {
    name: string;
    type: number;
}

const CustomersScreen:React.FC =()=> {
    const {listItems, isLoading} = useFetchUserData()
    const [filteredList, setFilteredList] = useState<User[]>([])

    useEffect(()=>{
        filterListByUserType('admin')
    },[listItems])

    const filterListByUserType = (userType: string)=>{
        const resultList = filterUsersByType(listItems, userType);
        setFilteredList(resultList);
    }

    const handleUserChange=(userType:string)=>{
        filterListByUserType(userType)
    }

    if (isLoading){
        return(<View style={styles.loaderView}>
          <ActivityIndicator testID='loader' color={colors.loader} size='small'/>
        </View>) 
      }

    return (
      <ScrollView style={styles.container} showsVerticalScrollIndicator={false} testID="customers-screen">
        <CategoryHeader category='User Types'/>
        <CustomRadioButton handleSelectedUserType={handleUserChange}/>
        <LineSeparator />
        <CategoryHeader category='Manager Users'/>
        {filteredList.length>0 && filteredList.map((item: User, index: number)=><UserListItem name={item.name} userType={item.type} key={index} index={index} />)}
        <LineSeparator />
      </ScrollView>
    );
}

const styles = StyleSheet.create({
    container: {
      flex: 1,
      paddingHorizontal:24,
      backgroundColor: colors.primary
    },
    loaderView:{
      flex:1, 
      justifyContent:'center', 
      alignItems:'center'
    }
});  

export default CustomersScreen
