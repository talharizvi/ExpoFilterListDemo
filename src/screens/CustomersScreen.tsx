import React, { useState, useEffect, useCallback} from 'react';
import { StatusBar } from 'expo-status-bar';
import { StyleSheet, Text, View, ScrollView, ActivityIndicator } from 'react-native';
import { useFetchUserData } from '../hooks/useFetchUserData';
import UserListItem from '../components/UserListItem';
import CategoryHeader from '../components/CategoryHeader';
import CustomRadioButton from '../components/CustomRadioButton';
import LineSeparator from '../components/LineSeparator';
import colors from '../constants/colors';
import { filterUsersByType } from '../utils/utils';

interface User {
    name: string;
    type: number;
}

const CustomersScreen:React.FC =()=> {
    const {listItems, isLoading} = useFetchUserData()
    const [filteredList, setFilteredList] = useState<User[]>([])

    useEffect(()=>{
        filterListByUserType('admin')
    },[listItems])

    const filterListByUserType = (userType: string)=>{
        const resultList = filterUsersByType(listItems, userType);
        setFilteredList(resultList);
    }

    const handleUserChange=(userType:string)=>{
        filterListByUserType(userType)
    }

    if (isLoading){
        return(<View style={styles.loaderView}>
          <ActivityIndicator color={colors.loader} size='small'/>
        </View>) 
      }

    return (
      <ScrollView style={styles.container} showsVerticalScrollIndicator={false}>
        <CategoryHeader category='User Types'/>
        <CustomRadioButton handleSelectedUserType={handleUserChange}/>
        <LineSeparator />
        <CategoryHeader category='Manager Users'/>
        {filteredList.map((item: User, index: number)=><UserListItem name={item.name} userType={item.type} key={index}/>)}
        <LineSeparator />
      </ScrollView>
    );
}

const styles = StyleSheet.create({
    container: {
      flex: 1,
      paddingHorizontal:24,
      backgroundColor: colors.primary
    },
    loaderView:{
      flex:1, 
      justifyContent:'center', 
      alignItems:'center'
    }
});  

export default CustomersScreen
