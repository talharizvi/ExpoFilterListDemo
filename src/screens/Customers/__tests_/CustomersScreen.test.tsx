import React from 'react';
import { render, waitFor, screen, fireEvent } from '@testing-library/react-native';
import CustomersScreen from '../CustomersScreen';
import { useFetchUserData } from '../../../hooks/useFetchUserData';
import colors from '../../../constants/colors';
jest.mock('../../../hooks/useFetchUserData');

describe('CustomersScreen', () => {
  it('displays loader when isLoading is true', () => {
    (useFetchUserData as jest.Mock).mockReturnValue({ listItems: [], isLoading: true });

    const { toJSON, getByTestId } = render(<CustomersScreen />);
    expect(getByTestId('loader')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });
});

it('does not display loader when isLoading is false', () => {
  (useFetchUserData as jest.Mock).mockReturnValue({ listItems: [], isLoading: false });

  const { toJSON, queryByTestId } = render(<CustomersScreen />);
  expect(toJSON()).toMatchSnapshot();
  expect(queryByTestId('loader')).toBeNull();
});

test('displays user list when data is loaded', async () => {
  // Mocking listItems with sample data
  const mockData = [
    {
      "name": "Karan Ahuja",
      "type": 1
    },
    {
      "name": "Suraj Singh",
      "type": 0
    }
  ];
  jest.spyOn(require('../../../hooks/useFetchUserData'), 'useFetchUserData').mockReturnValueOnce({ listItems: mockData, isLoading: false });

  const { toJSON, getByTestId, queryByTestId } = render(<CustomersScreen />);

  // Check if the loader is not present
  await waitFor(() => expect(queryByTestId('loader')).toBeNull());

  // Check if the user list is rendered
  const userList = getByTestId('customers-screen');
  expect(userList).toBeTruthy();

  const adminButton = getByTestId('admin-button');
  expect(adminButton.props.style.backgroundColor).toBe(colors.background);

  await waitFor(() => {
    // Check if the user list items are rendered correctly after filtering
    expect(queryByTestId('user-0')).toBeTruthy();
    expect(queryByTestId('user-1')).toBeNull();
  });

  expect(toJSON()).toMatchSnapshot(); 
});