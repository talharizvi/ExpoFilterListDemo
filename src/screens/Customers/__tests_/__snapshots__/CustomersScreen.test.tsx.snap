// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomersScreen displays loader when isLoading is true 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    color="#007bff"
    size="small"
    testID="loader"
  />
</View>
`;

exports[`displays user list when data is loaded 1`] = `
<RCTScrollView
  showsVerticalScrollIndicator={false}
  style={
    {
      "backgroundColor": "#ffffff",
      "flex": 1,
      "paddingHorizontal": 24,
    }
  }
  testID="customers-screen"
>
  <View>
    <View
      testID="category-header"
    >
      <Text
        style={
          {
            "fontSize": 18,
            "fontWeight": "bold",
            "marginBottom": 8,
            "marginTop": 16,
          }
        }
      >
        User Types
      </Text>
    </View>
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 122, 255, 0.12)",
            "borderRadius": 8,
            "flexDirection": "row",
            "marginRight": 20,
            "marginTop": 10,
            "opacity": 1,
            "paddingLeft": 8,
            "paddingVertical": 12,
          }
        }
        testID="admin-button"
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderColor": "#dadee6",
              "borderRadius": 12,
              "borderWidth": 2,
              "height": 18,
              "justifyContent": "center",
              "width": 18,
            }
          }
        >
          <View
            style={
              {
                "backgroundColor": "#000000",
                "borderRadius": 3,
                "height": 6,
                "width": 6,
              }
            }
          />
        </View>
        <Text
          style={
            {
              "marginLeft": 18,
            }
          }
        >
          Admin
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "marginRight": 20,
            "marginTop": 10,
            "opacity": 1,
            "paddingLeft": 8,
            "paddingVertical": 12,
          }
        }
        testID="manager-button"
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderColor": "#dadee6",
              "borderRadius": 12,
              "borderWidth": 2,
              "height": 18,
              "justifyContent": "center",
              "width": 18,
            }
          }
        />
        <Text
          style={
            {
              "marginLeft": 18,
            }
          }
        >
          Manager
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#dadee6",
          "height": 1,
          "marginLeft": 2,
          "marginRight": 20,
          "marginTop": 18,
        }
      }
      testID="line-separator"
    />
    <View
      testID="category-header"
    >
      <Text
        style={
          {
            "fontSize": 18,
            "fontWeight": "bold",
            "marginBottom": 8,
            "marginTop": 16,
          }
        }
      >
        Manager Users
      </Text>
    </View>
    <View
      style={
        {
          "marginBottom": 24,
          "marginTop": 8,
        }
      }
      testID="user-0"
    >
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "rgba(0, 122, 255, 0.12)",
              "borderRadius": 8,
              "justifyContent": "center",
              "marginRight": 14,
              "paddingHorizontal": 12,
            }
          }
        >
          <Text
            style={
              {
                "color": "#5280eb",
              }
            }
          >
            S
          </Text>
        </View>
        <View>
          <Text
            style={
              {
                "fontSize": 14,
                "fontWeight": "600",
              }
            }
          >
            Suraj Singh
          </Text>
          <Text
            style={
              {
                "color": "#5e5e5d",
              }
            }
          >
            Admin
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#dadee6",
          "height": 1,
          "marginLeft": 2,
          "marginRight": 20,
          "marginTop": 18,
        }
      }
      testID="line-separator"
    />
  </View>
</RCTScrollView>
`;

exports[`does not display loader when isLoading is false 1`] = `
<RCTScrollView
  showsVerticalScrollIndicator={false}
  style={
    {
      "backgroundColor": "#ffffff",
      "flex": 1,
      "paddingHorizontal": 24,
    }
  }
  testID="customers-screen"
>
  <View>
    <View
      testID="category-header"
    >
      <Text
        style={
          {
            "fontSize": 18,
            "fontWeight": "bold",
            "marginBottom": 8,
            "marginTop": 16,
          }
        }
      >
        User Types
      </Text>
    </View>
    <View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 122, 255, 0.12)",
            "borderRadius": 8,
            "flexDirection": "row",
            "marginRight": 20,
            "marginTop": 10,
            "opacity": 1,
            "paddingLeft": 8,
            "paddingVertical": 12,
          }
        }
        testID="admin-button"
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderColor": "#dadee6",
              "borderRadius": 12,
              "borderWidth": 2,
              "height": 18,
              "justifyContent": "center",
              "width": 18,
            }
          }
        >
          <View
            style={
              {
                "backgroundColor": "#000000",
                "borderRadius": 3,
                "height": 6,
                "width": 6,
              }
            }
          />
        </View>
        <Text
          style={
            {
              "marginLeft": 18,
            }
          }
        >
          Admin
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "marginRight": 20,
            "marginTop": 10,
            "opacity": 1,
            "paddingLeft": 8,
            "paddingVertical": 12,
          }
        }
        testID="manager-button"
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderColor": "#dadee6",
              "borderRadius": 12,
              "borderWidth": 2,
              "height": 18,
              "justifyContent": "center",
              "width": 18,
            }
          }
        />
        <Text
          style={
            {
              "marginLeft": 18,
            }
          }
        >
          Manager
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#dadee6",
          "height": 1,
          "marginLeft": 2,
          "marginRight": 20,
          "marginTop": 18,
        }
      }
      testID="line-separator"
    />
    <View
      testID="category-header"
    >
      <Text
        style={
          {
            "fontSize": 18,
            "fontWeight": "bold",
            "marginBottom": 8,
            "marginTop": 16,
          }
        }
      >
        Manager Users
      </Text>
    </View>
    <View
      style={
        {
          "backgroundColor": "#dadee6",
          "height": 1,
          "marginLeft": 2,
          "marginRight": 20,
          "marginTop": 18,
        }
      }
      testID="line-separator"
    />
  </View>
</RCTScrollView>
`;
